In [2]:
%matplotlib inline


What is PyTorch?
================

It’s a Python-based scientific computing package targeted at two sets of
audiences:

-  A replacement for NumPy to use the power of GPUs
-  a deep learning research platform that provides maximum flexibility
   and speed

Getting Started
---------------

**Tensors**


Tensors are similar to NumPy’s ndarrays, with the addition being that
Tensors can also be used on a GPU to accelerate computing.



In [3]:
from __future__ import print_function
import torch

*Construct a 5x3 matrix, uninitialized:*



In [4]:
x = torch.empty(5, 3)
print(x)

tensor([[1.1210e-44, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00]])


*Construct a randomly initialized matrix:*



In [5]:
x = torch.rand(5, 3)
print(x)

tensor([[0.6660, 0.5067, 0.0620],
        [0.4552, 0.1673, 0.8351],
        [0.9707, 0.4417, 0.5752],
        [0.4494, 0.6565, 0.3179],
        [0.8163, 0.6865, 0.3603]])


*Construct a matrix filled zeros and of dtype long:*



In [6]:
x = torch.zeros(5, 3, dtype=torch.long)
print(x)

tensor([[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]])


*Construct a tensor directly from data:*



In [7]:
x = torch.tensor([5.5, 3])
print(x)

tensor([5.5000, 3.0000])


*Get its size:*



In [8]:
print(x.size())

torch.Size([2])


**Tensor Operations**

*There are multiple syntaxes for operations. In the following example, we look at the addition operation*

*Addition syntax 1*

In [10]:
y= torch.rand(5, 3)
x= torch.rand(5 ,3)
print(x + y)

tensor([[1.0563, 0.9405, 0.7861],
        [1.3442, 0.2175, 0.4697],
        [0.9339, 0.9745, 1.0829],
        [1.4741, 0.4026, 1.6240],
        [1.4712, 0.8823, 1.4359]])


*Addition: syntax 2*



In [11]:
print(torch.add(x, y))

tensor([[1.0563, 0.9405, 0.7861],
        [1.3442, 0.2175, 0.4697],
        [0.9339, 0.9745, 1.0829],
        [1.4741, 0.4026, 1.6240],
        [1.4712, 0.8823, 1.4359]])


Addition: providing an output tensor as argument



In [12]:
result = torch.empty(5, 3)
torch.add(x, y, out=result)
print(result)

tensor([[1.0563, 0.9405, 0.7861],
        [1.3442, 0.2175, 0.4697],
        [0.9339, 0.9745, 1.0829],
        [1.4741, 0.4026, 1.6240],
        [1.4712, 0.8823, 1.4359]])


*You can use standard NumPy-like indexing! Remember that python indexing is [x,y] which is opposite to matlab's (y,x)!*

In [13]:
print(x[:, 1])

tensor([0.5989, 0.0088, 0.3546, 0.0163, 0.6292])


*Resizing: If you want to resize/reshape tensor, you can use ``torch.view``:*



In [14]:
x = torch.randn(4, 4)
y = x.view(16)
z = x.view(-1, 8)  # the size -1 is inferred from other dimensions
print(x.size(), y.size(), z.size())

torch.Size([4, 4]) torch.Size([16]) torch.Size([2, 8])


*If you have a one element tensor, use ``.item()`` to get the value as a
Python number*



In [15]:
x = torch.randn(1)
print(x)
print(x.item())

tensor([-1.9483])
-1.9482558965682983


**Read later:**


  100+ Tensor operations, including transposing, indexing, slicing,
  mathematical operations, linear algebra, random numbers, etc.,
  are described
  `here <https://pytorch.org/docs/torch>`_.

NumPy Bridge
------------

Converting a Torch Tensor to a NumPy array and vice versa is a breeze.


**Converting a Torch Tensor to a NumPy Array**


In [20]:
a = torch.ones(5)
print(a)

tensor([1., 1., 1., 1., 1.])


In [21]:
b = a.numpy()
print(b)

[1. 1. 1. 1. 1.]


**Converting NumPy Array to Torch Tensor**

*See how changing the np array changed the Torch Tensor automatically*


In [23]:
import numpy as np
a = np.ones(5)
b = torch.from_numpy(a)
np.add(a, 1, out=a)
print(a)
print(b)

[2. 2. 2. 2. 2.]
tensor([2., 2., 2., 2., 2.], dtype=torch.float64)




CUDA Tensors
------------

Tensors can be moved onto any device (GPU or CPU) using the ``.to`` method.



In [24]:
# let us run this cell only if CUDA is available
# We will use ``torch.device`` objects to move tensors in and out of GPU
if torch.cuda.is_available():
    device = torch.device("cuda")          # a CUDA device object
    y = torch.ones_like(x, device=device)  # directly create a tensor on GPU
    x = x.to(device)                       # or just use strings ``.to("cuda")``
    z = x + y
    print(z)
    print(z.to("cpu", torch.double))       # ``.to`` can also change dtype together!



Autograd: Automatic Differentiation
===================================

Central to all neural networks in PyTorch is the ``autograd`` package. When building a neural network
with pytorch operations, we are specifying the "feed-forward" operations of the neural network. 
For back-propgation, the autograd package provides automatic differentiaion for all tensor operations 
specified in the network

Let’s first briefly visit this, and we will then go to training our
first neural network.



Tensor
--------

``torch.Tensor`` is the central class of the package. If you set its attribute
``.requires_grad`` as ``True``, it starts to track all operations on it. When
you finish your computation you can call ``.backward()`` and have all the
gradients computed automatically. The gradient for this tensor will be
accumulated into ``.grad`` attribute.

To stop a tensor from tracking history, you can call ``.detach()`` to detach
it from the computation history, and to prevent future computation from being
tracked.

To prevent tracking history (and using memory), you can also wrap the code block
in ``with torch.no_grad():``. This can be particularly helpful when evaluating a
model because the model may have trainable parameters with
``requires_grad=True``, but for which we don't need the gradients.

There’s one more class which is very important for autograd
implementation - a ``Function``.

``Tensor`` and ``Function`` are interconnected and build up an acyclic
graph, that encodes a complete history of computation. Each tensor has
a ``.grad_fn`` attribute that references a ``Function`` that has created
the ``Tensor`` (except for Tensors created by the user - their
``grad_fn is None``).

If you want to compute the derivatives, you can call ``.backward()`` on
a ``Tensor``. If ``Tensor`` is a scalar (i.e. it holds a one element
data), you don’t need to specify any arguments to ``backward()``,
however if it has more elements, you need to specify a ``gradient``
argument that is a tensor of matching shape.


*Create a tensor and set ``requires grad = true`` to track computation with it*

In [28]:
x = torch.ones(2,2, requires_grad = True)
print(x)

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)


*Do a tensor operation*

In [29]:
y = x + 2
print(y)

tensor([[3., 3.],
        [3., 3.]], grad_fn=<AddBackward0>)


*`y` was created as a result of an operation, so it has a grad_fn*

In [31]:
print(y.grad_fn)

*Do more operations on ``y``*



In [32]:
z = y * y * 3
out = z.mean()

print(z, out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>) tensor(27., grad_fn=<MeanBackward0>)


``.requires_grad_( ... )`` *changes an existing Tensor's ``requires_grad``
flag in-place. The input flag defaults to ``False`` if not given.*



In [33]:
a = torch.randn(2, 2)
a = ((a * 3) / (a - 1))
print(a.requires_grad)
a.requires_grad_(True)
print(a.requires_grad)
b = (a * a).sum()
print(b.grad_fn)


False
True


Gradients
---------
Let's backprop now.
Because ``out`` contains a single scalar, ``out.backward()`` is
equivalent to ``out.backward(torch.tensor(1.))``.



In [34]:
out.backward()

Print gradients d(out)/dx


In [35]:
print(x.grad)

tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


You should have got a matrix of ``4.5``. Let’s call the ``out``
*Tensor* “$o$”.
We have that $o = \frac{1}{4}\sum_i z_i$,
$z_i = 3(x_i+2)^2$ and $z_i\bigr\rvert_{x_i=1} = 27$.
Therefore,
$\frac{\partial o}{\partial x_i} = \frac{3}{2}(x_i+2)$, hence
$\frac{\partial o}{\partial x_i}\bigr\rvert_{x_i=1} = \frac{9}{2} = 4.5$.

